File: akka_essentials/src/main/scala/essentials/actor/scala/chapter2/actor/ReduceActor.scala
package essentials.actor.scala.chapter2.actor

import akka.actor.Actor
import essentials.actor.scala.chapter2.{MapData, ReduceData, WordCount}


class ReduceActor extends Actor {

  def receive: Receive = {
    case MapData(dataList) => sender ! reduce(dataList)
  }

  def reduce(words: IndexedSeq[WordCount]): ReduceData = ReduceData {
    words.foldLeft(Map.empty[String, Int]) { (index, words) =>
      index + (words.word -> (index.getOrElse(words.word, 1) + 1))
    }
  }

}

File: akka_essentials/src/main/java/essentials/actor/java/chapter2/actor/ReduceActor.java
package essentials.actor.java.chapter2.actor;

import akka.actor.UntypedActor;
import essentials.actor.java.chapter2.message.MapData;
import essentials.actor.java.chapter2.message.ReduceData;
import essentials.actor.java.chapter2.message.WordCount;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class ReduceActor extends UntypedActor {

    @Override
    public void onReceive(Object message) throws Exception {
        if (message instanceof MapData) {
            MapData mapData = (MapData) message;
            // reduce the incoming data and forward the result to Master actor
            sender().tell(reduce(mapData.getDataList()), self());
        } else
            unhandled(message);
    }

    private ReduceData reduce(List<WordCount> dataList) {
        Map<String, Integer> reducedMap = new HashMap<>();
        for (WordCount wordCount : dataList) {
            Integer value = reducedMap.getOrDefault(wordCount.getWord(), 0);
            value++;
            reducedMap.put(wordCount.getWord(), value);
        }
        return new ReduceData(reducedMap);
    }

}
